spec/units/clever_tap_client_spec.rb

```diff
@@ -38,4 +39,272 @@
+            it_behaves_like 'it yields control', 'once'
+          end
+        end
+
+        context 'and objects do not fit `upload_limit`' do
+          before do
+            allow(CleverTap::Event)
+              .to receive(:upload_limit).and_return(1)
+          end
+
+          VCR.use_cassette('upload out of Event limit') do
+            it_behaves_like 'it calls success hook', 'twice'
+            it_behaves_like 'it yields control', 'twice'
+          end
+        end
+      end
+
+      context 'and invalid records' do
+        let(:failure_proc) { proc { 'sample proc' } }
+
+        subject do
+          client = described_class.new
+          client.on_failed_upload(&failure_proc)
+          client.send(:batched_upload, CleverTap::Profile, [profile], 0)
+        end
+
+        let(:profile) { profile_factory }
+
+        it_behaves_like 'it calls failure hook', 'once'
+      end
+    end
+
+    context 'when objects are not homogenous' do
+      let(:event1) { event_factory('1414', 'John') }
+      let(:event2) { {} }
+
+      it_behaves_like 'it raises an inconcistency error'
+    end
+  end
+
+  describe '#upload' do
+    let(:success_proc) { proc { 'sample proc' } }
+
+    subject do
+      client = described_class.new
+      client.on_successful_upload(&success_proc)
+      client.upload([event1, event2])
+    end
+
+    before { setup_clevertap_client }
+
+    context 'when upload records are homogenous' do
+      context 'and valid records' do
+        let(:event1) { event_factory('1414', 'John') }
+        let(:event2) { event_factory('1515', 'Jill') }
+
+        context 'and objects fit `upload_limit`' do
+          VCR.use_cassette('upload within Event limit') do
+            it 'returns an array with one successful `Response` object' do
+              expect(subject.count).to eq 1
+              expect(subject.first.success).to be true
+              expect(subject.first.class).to eq CleverTap::Response
+            end
+
+            it_behaves_like 'it calls success hook', 'once'
+          end
+        end
+
+        context 'and objects do not fit `upload_limit`' do
+          before do
+            allow(CleverTap::Event)
+              .to receive(:upload_limit).and_return(1)
+          end
+
+          VCR.use_cassette('upload out of Event limit') do
+            it 'returns an array with two successful `Response` objects' do
+              expect(subject.count).to eq 2
+              expect(subject.all?(&:success)).to be true
+              expect(
+                subject.all? { |r| r.class == CleverTap::Response }
+              ).to be true
+            end
+
+            it_behaves_like 'it calls success hook', 'twice'
+          end
+        end
+      end
+
+      context 'and invalid records' do
+        let(:failure_proc) { proc { 'sample proc' } }
+
+        subject do
+          client = described_class.new
+          client.on_failed_upload(&failure_proc)
+          client.upload([profile])
```

Comment:
it's weird to set the subject as a result of a method.

Reply:
That's the only good way (I managed to think of at the time) to abstract the shared examples for the two tested methods (that do very similar things but with different parameters passed)